File: Collections/Edge.cs
using System;

namespace Collections
{
    /// <summary>
    /// Weighted directed graph edge.
    /// </summary>
    public class Edge<Vertex> : IComparable<Edge<Vertex>>
    {
        public Vertex from;
        public Vertex to;
        public double weight;

        public int CompareTo(Edge<Vertex> edge) => weight.CompareTo(edge.weight);

        public override string ToString() => $"{from} -> {to}";

        public Edge<Vertex> Complement() => new Edge<Vertex> { from = to, to = from, weight = weight };
    }
}

File: src/main/java/in/bfdes/collections/Edge.java
package in.bfdes.collections;

public record Edge<V>(V from, V to, double weight) implements Comparable<Edge<V>> {
    public Edge<V> complement() {
        return new Edge<>(to, from, weight);
    }

    @Override
    public int compareTo(Edge<V> edge) {
        return Double.compare(weight, edge.weight);
    }

    @Override
    public String toString() {
        return String.format("%s -( %f )-> %s", from, weight, to);
    }
}
